Add "with-design-picker" onboarding flow

ciampo opened the pull request:
## Changes proposed in this Pull Request

* Create a new signup step (`design`) and a new signup flow (`onboarding-with-design`)
* Add a new component responsible for rendering the `design` signup step, which is basically a wrapper around the `<DesignPicker />` component from the `@automattic/design-picker` component
* Update the parameters (sent to the backend) used to create a new site for when there's a selected design

**Caveats:**

-  theming the `<DesignPicker />` component is tracked separately in #51401 and is not a goal of this PR
- the `Uncaught (in promise) TypeError: Cannot read property 'replace' of undefined at eval (controller.js:308)` console error is also reproducible in `trunk` and therefore is not introduced by this PR

## Testing instructions

### Make sure that all A/Cs are fulfilled

A/Cs are defined in #51230

### Test new flow (free plan, selecting a design)

- Visit `/start/with-design-picker`
- Pick a free domain
- Pick the free plan
- Pick a design
- Check that:
  - [ ] Site is created successfully
  - [ ] Site is assigned the design that was picked during the signup flow

### Test new flow (paid plan, selecting a design)\

- Visit `/start/with-design-picker`
- Pick a free domain
- Pick a paid plan
- Pick a design
- Check that:
  - [ ] Browser is redirected to checkout
  - [ ] Upon completion of the checkout, the site is created successfully
  - [ ] Site is assigned the design that was picked during the signup flow

### Test new flow (free plan, skipping design)

- Visit `/start/with-design-picker`
- Pick a free domain
- Pick the free plan
- Skip the design step (scroll all the way to the bottom of the screen and click on ignore link)
- Check that:
  - [ ] Site is created successfully
  - [ ] Site is assigned the default design (Herve)

### Test new flow (change locale, free plan, selecting a design)

- In the WordPress.com settings, change your language to a locale different than english
- Visit `/start/with-design-picker`
- Pick a free domain
- Pick the free plan
- When on the design step, check that:
  - [ ] All strings (title, subtitle, ignore link) are translated

### Smoke tests other flows 

- Visit `/start` or any other signup flow
- Smoke test:
  - [ ] Everything should keep working as expected




Closes #51230

ciampo committed:
Add DesignPickerStep component

ciampo committed:
Define "design" onboarding step

ciampo committed:
Add 'with-design-picker' onboarding flow (including selectedDesign schema and serialisation)

ciampo committed:
Include design picker options to new site params

ciampo committed:
Update flow 'lastModified' param

ciampo commented on client/signup/config/flows-pure.js:
```diff
@@ -437,6 +437,14 @@ export function generateFlows( {
 		showRecaptcha: true,
 	};
 
+	flows[ 'with-design-picker' ] = {
```
Note to self: this is how a new flow is defined with the steps that the user will go through

ciampo commented on client/signup/config/step-components.js:
```diff
@@ -71,6 +71,7 @@ const stepNameToModuleName = {
 	'plans-ecommerce-monthly': 'plans',
 	'plans-personal-monthly': 'plans',
 	'plans-premium-monthly': 'plans',
+	design: 'design-picker',
```
Note to self: this is where an onboarding step is associated to a React component (by convention, defined in the `signup/steps/[SLUG]/` folder)

In this case, the `design` step is associated to the component defined in ``signup/steps/design-picker/index.jsx`

ciampo commented on client/signup/config/steps-pure.js:
```diff
@@ -721,6 +721,12 @@ export function generateSteps( {
 			},
 		},
+
+		design: {
```
Note to self: this is how an onboarding step is defined. This step can then be used in onboarding flows ([example](https://github.com/Automattic/wp-calypso/pull/51501/files#r603335613))
The `stepName` will define the URL that the user will see in their browser when on this step.

The definition of a step, and [the component that a step renders](https://github.com/Automattic/wp-calypso/pull/51501/files#r603337015), are decoupled — this is because one React component can be used by more than one step (e.g. with different props / dependencies)

a8ci18n commented on client/signup/steps/design-picker/index.jsx:
```diff
@@ -0,0 +1,89 @@
+	subHeaderText() {
+		const { translate } = this.props;
+
+		return translate( 'Pick your favorite homepage layout. You can customize or change it later.', {
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **20** times:
`translate( 'Pick your favorite homepage layout. You can customize or change it later.' )` *ES Score: 10*


:information_source: This string is probably long enough that it doesn't need a **context**. Are you sure you don't want to use a translator comment instead?

ciampo replied:
@razvanpapadopol 🤷 

Looks like adding context to the translation triggers a new translation (instead of reusing the old one) — should we keep the context and wait for the new translated strings?

razvanpapadopol replied:
Yes, if the new `context` value doesn't match a previously entered value, it's basically a new translation. But do we really need a different way to translate the original string in this case?

ciampo replied:
I don't think we need a new translation with added context — the "old" translation is used exactly for the same purpose in [gutenboarding](https://github.com/Automattic/wp-calypso/blob/c9fe2bd56261aff7454a828785c74be9a463d17f/client/landing/gutenboarding/onboarding-block/design-selector/index.tsx#L53).
I also checked a few translations of this string without context, and they all look good.

I'm going to remove the context

ciampo committed:
Add empty styles for <DesignPickerStep />

ciampo commented on client/state/signup/progress/actions.js:
```diff
@@ -64,6 +64,10 @@ function recordSubmitStep( stepName, providedDependencies ) {
 			}
 
+			if ( includes( [ 'selected_design' ], propName ) ) {
+				propValue = propValue.slug;
```
@razvanpapadopol is there a reason why, in your draft PR, you didn't pass the `selectedDesign.slug` in [`this.props.submitSignupStep`](https://github.com/Automattic/wp-calypso/pull/51501/files#diff-425a2aeab314234f95622905026eef2b94eeafa1ecbd493c5bf3d519d44757afR42-R50), instead of transforming it after?

Maybe I'm missing something

razvanpapadopol replied:
I was thinking that we need to pass `selectedDesign` object in `providedDependencies` for its values to be available in `client/lib/signup/step-actions/index.js` but if there is a way to pass just the slug for recording purposes, that's fine.

ciampo replied:
Looks like you're right. the `selectedDesign` dependency needs to provided in full (ie. the full object) for the other step actions to work (e.g. adding new site parameters)

ciampo committed:
Import DesignPicker from `@automattic/design-picker` package

ciampo committed:
Use correct value for the locale passed to <DomainPicker />

ciampo commented on client/lib/signup/step-actions/index.js:
```diff
@@ -174,6 +174,24 @@ function getNewSiteParams( {
+	// in gutenboarding (see `client/landing/gutenboarding/stores/onboard/actions.ts`)
+	const selectedDesign = get( signupDependencies, 'selectedDesign', false );
+	const withDesignOptions = {
+		theme: `pub/${ selectedDesign?.theme || themeSlugWithRepo }`,
```
Not entirely sure if, in case `selectedDesign.?theme` is falsey, we should prepend `pub/` also to `themeSlugWithRepo`.

Also, is there a reason why we wouldn't use `theme` instead of `themeSlugWithRepo` ?

razvanpapadopol replied:
Probably we need to rethink the creation of the `withDesignOptions` object since [we use it](https://github.com/Automattic/wp-calypso/pull/51501/files#diff-a621c4ea73ecba2ef5f903f325105824ebd61dcc12da5f505d2e0031bc4a4fc9R214) only when `selectedDesign` is defined and, in that case, it always includes `theme`, `template` and `fonts` properties.

ciampo replied:
I partially reworked this logic in https://github.com/Automattic/wp-calypso/pull/51501/commits/70f217406b700d0db5e962d19f2bc95d466aab74 and https://github.com/Automattic/wp-calypso/pull/51501/commits/09c123090c56ccf20c99a60aba96eee2efa693dd

This should fit more the code style of this file, and gracefully fallback to using the default theme when `selectedDesign` is undefined

razvanpapadopol replied:
Looks good! Thanks!

ciampo commented on client/signup/steps/design-picker/index.jsx:
```diff
@@ -0,0 +1,91 @@
+
+	renderDesignPicker() {
+		// props.locale obtained via `localize` HoC
+		return <DesignPicker locale={ this.props.locale } onSelect={ this.pickDesign } />;
```
In order to achieve feature parity, should the designs be randomised?

Currently, in Gutenboarding designs are shuffled and saved in the `onboard` store (see `client/landing/gutenboarding/index.tsx`)

ciampo replied:
Not part of this PR, but need to open a new issue to track it before merging

razvanpapadopol replied:
We may do that in the future. We'll have the randomization function in `design-picker` package soon. See https://github.com/Automattic/wp-calypso/issues/51404

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5628591

Hi @ciampo, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include this string: `Pick your favorite homepage layout. You can customize or change it later.`

Thank you in advance!

ciampo committed:
Remove unnecessary context from design step subheader translation

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5628591).

ciampo committed:
Make selectedDesign optional, in case the user skips the design step

ciampo committed:
Refactor how extra new site params are added with a selectedDesign

- Assume that selectedDesign is defined, all of its props are defined too
- Move expansion of newSiteParams at the bottom of the function, to match existing code
- This means that the default theme works as in every other flow

ciampo commented on client/lib/signup/step-actions/index.js:
```diff
@@ -227,6 +228,23 @@ function getNewSiteParams( {
+		//(see https://github.com/Automattic/wp-calypso/issues/51231)
+		// Is this override necessary? Before overriding, the site_creation_flow
+		// is "with-design-picker"
+		newSiteParams.options.site_creation_flow = 'onboarding-with-design';
```
Wondering if this override is necessary? Before overriding, the site_creation_flow is `with-design-picker`

razvanpapadopol replied:
Good point! We should probably remove the override and use the same flow value regardless of the design selection (including no selection/ skip). Then we should take this into consideration when implementing https://github.com/Automattic/wp-calypso/issues/51231 so a _Gutenboarding-like_ site to be created only when `template` and `fonts` are passed as well 🤔

ciampo replied:
Backend check in #51231 could become:

`if (isGutenboarding || isSiteWithThemeFontsAndBlocks()) { ... }`

ciampo committed:
Do not override site_creation_flow

ciampo committed:
No need to add `selectedDesign` in the step argument

ciampo committed:
Temporarily override site creation flow to `gutenboarding`

This will allow us to test that the site is created with all the correct
parameters (design, theme, fonts, headstart content...)

ciampo committed:
Filter out anchorfm designs by default

ciampo committed:
Add horizontal spacing around design step content

alshakero committed:
Don't fake the site_creation_flow value

ciampo commented on client/lib/signup/step-actions/index.js:
```diff
@@ -227,6 +228,15 @@ function getNewSiteParams( {
 	}
 
+	if ( selectedDesign ) {
+		// If there's a selected design, it means we're in "with_design_picker" flow.
```
This comment could easily become outdated if we introduce other flows with the `design` step — what if we changed it to something like

```suggestion
		// If there's a selected design, it means that the current flow contains the "design" step.
```

Commit:
Update client/lib/signup/step-actions/index.js

Co-authored-by: Marco Ciampini <marco.ciampini@automattic.com>

alshakero commented on packages/design-picker/src/components/index.tsx:
```diff
@@ -34,7 +34,10 @@ interface Props {
-	designs = availableDesigns.featured,
+	designs = availableDesigns.featured.filter(
+		// By default, exclude anchorfm-specific designs
+		( design ) => design.features.findIndex( ( f ) => f === 'anchorfm' ) < 0
```
Nit: `includes` serves this purpose more accurately.

ciampo replied:
That's what I tried to write in the first place, but it seems like our build configuration doesn't allow for it — I get an error saying

```Property 'includes' does not exist on type '"anchorfm"[]'. Do you need to change your target library? Try changing the `lib` compiler option to 'es2016' or later.```

I also tried to use it on `Array.prototype` and I got the same error

razvanpapadopol replied:
We have examples of packages using [ES2017](https://github.com/Automattic/wp-calypso/blob/trunk/packages/composite-checkout/tsconfig.json#L4) or [ESNext](https://github.com/Automattic/wp-calypso/blob/trunk/packages/language-picker/tsconfig.json#L4) as `lib` option.

ciampo replied:
Great to hear! Definitely going to switch to that for next time

alshakero reviewed (approved):
Tested every scenario and reviewed the code. All works as it should.

ciampo commented:
> Tested every scenario and reviewed the code. All works as it should.

Thank you for reviewing! @razvanpapadopol Is this PR good to be merged?
